### ISSUE TITLE
Visualization engine in theano like tensorboard

### ISSUE THREAD

**Issue by shashankg7** [Theano/Theano#4909]
Is there any visualization engine which can be integrated with theano to monitor training like tensorboard in tensorflow?

It's hard to visualize what is going on during training by looking at raw numbers.

**Comment by nouiz** [comment]
There isn't something as nice as tensorboard.

There is 3dviz to visualize Theano function:

http://deeplearning.net/software/theano/library/d3viz/index.html

The doc about other printing in Theano:

http://deeplearning.net/software/theano/library/printing.html?highlight=printing

Some framework on top of Theano have there own way to monitor during
training what is happening. Do you use a framework on top of Theano? You
can look at Keras, Lasagne and blocks. I don't know how there ploting work,
but I know at least 2 of thems have that.

**Comment by shashankg7** [comment]
@nouiz Thanks for replying. For the current project I am using plain theano, hence the question. 

As far as I know keras supports visualization through tensorboard only. I am not sure if that would work with theano backend. And I don't know about visualization in blocks and lasagne.

**Comment by kariannbriski** [comment]
Have you looked at DIGITS? It does not support Theano at this time. https://developer.nvidia.com/digits

**Comment by botev** [comment]
On this topic, since Pytorch has integration with Tensorboard, maybe it is possible to create similar integration with Theano as well, such that rather than having to build a whole visualization library, which could be quite a lot of work, just build a middle layer for dumping Theano variables in the correct format to a folder and visualize in Tensorboard?

**Comment by nouiz** [comment]
I know some people use tensorboard with Theano. So there is code existing
for this.

Le dim. 28 mai 2017 07:39, Alexander Botev <notifications@github.com> a
écrit :

**Comment by botev** [comment]
@nouiz  Could you point me to any repositories doing this? 
So far I found these:
https://github.com/bravomikekilo/mxconsole - this is suppose to be platform neutral
https://github.com/alexguo/tensorboard_logger 
https://github.com/dmlc/tensorboard - an MXNet port to tensorboard

Would you guys be up to accepting Logging Ops in Theano for tensorboard if it is possible?

**Comment by lamblin** [comment]
Yes, integration with tensorboard would be really cool!

**Comment by nouiz** [comment]
I think we should accept Logging Ops if many people find that useful to have them in the graph instead of outside the graph.

There is also the new visdom project that could be interresting: https://github.com/facebookresearch/visdom

I didn't found link to what I had in mind. I think it was tensorboard_logger. There is also https://github.com/bartvm/mimir for logging. Now sure how it connect to tensorboard, but it was named in discussion related to tensorboard.

**Comment by botev** [comment]
@nouiz Does the fetching data gpu->cpu happen only at the end of the theano function or does it happen async during the graph computation as the rest of the ops? 

Whether it should be inside Theano is more dependable of how much you guys want this to be a convince, which I think makes sense, but is not my decision to make.

**Comment by botev** [comment]
Ok so maybe I've asked the question not too well.

The main "gain" to have the visualization (no matter if it is tensorboard or others like visdom) is that it can execute async, meaning that "summary(gpu)->summary(cpu)->summary(file)" can happen potentially while other parts of the graph are executing, without any performance hit. When I've asked about async I was meaning whether Theano waits for everything in order to return to the CPU the correct result. I think your answer means it doesn't so the above scenario is possible to make the summaries  without performance hit (although I don't know if you don't need a separate thread for writing to file).

The alternative, which is to write the summaries outside theano and return them, would mean that the user always will get a small hit due to the I/O of the summary and having to wait absolutely all and everything to be returned from the theano function.

Now, even if we have the visalization library, I'm not too sure how to wire up in Theano the transfers such that they don't block the overall graph execution. If that is possible we can make logging ops which require only python hooks and are independent of the visualization library.

**Comment by nouiz** [comment]
Current Theano cpu/gpu transfers are synchronous to make everything work
well.

But it is possible to make special ops that would do some transfer async.
I'm not sure how complicated it is. If the "hit" is 1%, it is probably not
worth the complexity of having it really async to don't have any hit. There
is other optimization that can give more then 1% speed up. If you have any
indication that the hit would be significant, then that is another story.

Also, depending of what summary is doing, maybe part of it could be
computed async on the GPU directly (like computing the min/max/mean) of
stuff you want to appear in the board.

**Comment by botev** [comment]
I do agree it depends how much % of the compute time the hit, is but for Variational Models, which need many iterations each could be quite fast, I just have this gut feeling that native python code with the I/O could be expensive. 

I will try installing tensorboard and visdom and see what are the in python timings for logging and will report back. And yes it would be best to compute the summaries inside Theano and directly transfer the final arrays only.

**Comment by botev** [comment]
@nouiz so it seems that it depends on how much data you log in it quite be a bit. However, note that you can offload the writing to a separate thread outside of theano and that makes the overhead less than 1%. 
I've tested this with the dmlc tensorboard. However it would be nice if Theano has maybe a few operators like histograms which to be performed inside theano rather than in numpy.

**Comment by lamblin** [comment]
I think `MultinomialFromUniform` (which has a GPU implementation) could be used to get histograms.

**Comment by nouiz** [comment]
Note, it would need to use the n paramter to do the reduction inside the
kernel. It is a pretty good code reuse!

**Comment by botev** [comment]
I'm not too sure I follow on that as I don't really know how people generate histograms. Could someone point me to a code or just show how you would make a histogram via Theano?
